### PR TITLE
Elm 4971 reverse mapping up3 to cemo

### PR DESCRIPTION
**Pull request by james-francis-MT**
### Context

Ticket: https://dsdmoj.atlassian.net/browse/ELM-4971

Probation can be editing a form that was created by a prison user a long time ago. 

Since it was created probation may have made a number of changes to that form offline by email.

When probation are onboarded and make a change to a form we need to get the latest version of the form from SERCO so that they are changing the actual order not a previous version of the order.

### Changes proposed in this pull request

Given I have made a call to SNOW

When the system gets the information back from SNOW

Then I need to reverse map from UP3 schema to CEMO domain model

# Backend PR Checklist

These are items (excluding GitHub Checks) which should be confirmed before a branch is ready to merge.

- [x] Ensure backwards compatibility (did not remove existing code, only added new)
- [x] Did not remove/edit existing enum values
- [x] Added relevant automation tests (updated or new)
- [x] Verified Serco Mapping changes are safe for production (e.g Postman)
- [x] Relevant documentation is updated and linked
- [x] PR has been self-reviewed by the author
- [x] Reused existing components before creating new ones
- [x] Code refined to the best of your knowledge
- [x] Ticket number included in the description
- [ ] (If applicable) Ran frontend scenario tests against backend changes

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/Offence.kt [hunk 1]
```diff
@@ -23,5 +23,9 @@ enum class Offence(val value: String) {
     fun from(value: String?): Offence? = Offence.entries.firstOrNull {
       it.name == value
     }
+
+    fun fromValue(value: String?): Offence? = Offence.entries.firstOrNull {
+      !value.isNullOrBlank() && it.value == value
```
Comment (CslBing) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/Offence.kt]: Shouldn't empty string map to NO_OFFENCE_COMMITTED? Or is this something cannot be determined?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/OrderTypeDescription.kt [hunk 2]
```diff
@@ -7,4 +7,11 @@ enum class OrderTypeDescription(val value: String?) {
   GPS_ACQUISITIVE_CRIME_HDC("GPS Acquisitive Crime HDC"),
   GPS_ACQUISITIVE_CRIME_PAROLE("GPS Acquisitive Crime Parole"),
   UNKNOWN(null),
```
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/OrderTypeDescription.kt:9]: same here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/Pilot.kt [hunk 3]
```diff
@@ -12,4 +12,9 @@ enum class Pilot(val value: String) {
   GPS_ACQUISITIVE_CRIME_HOME_DETENTION_CURFEW("GPS Acquisitive Crime Home Detention Curfew"),
   GPS_ACQUISITIVE_CRIME_PAROLE("GPS Acquisitive Crime Parole"),
   UNKNOWN(""),
```
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/Pilot.kt:14]: same here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/PoliceAreas.kt [hunk 4]
```diff
@@ -27,5 +27,9 @@ enum class PoliceAreas(val value: String) {
     fun from(value: String?): PoliceAreas? = PoliceAreas.entries.firstOrNull {
       it.name == value
     }
+
+    fun fromValue(value: String?): PoliceAreas? = PoliceAreas.entries.firstOrNull {
+      it.value == value
```
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/PoliceAreas.kt]: map to different if null?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/RiskCategory.kt [hunk 5]
```diff
@@ -18,4 +18,9 @@ enum class RiskCategory(val value: String) {
   HOMOPHOBIC_VIEWS("Evidence known to the Device Wearer having homophobic views"),
   UNDER_18("Under 18 living at property"),
   NO_RISK("No risk"),
+  ;
+
+  companion object {
+    fun from(value: String?): RiskCategory? = RiskCategory.entries.firstOrNull { it.value == value }
+  }
```
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/RiskCategory.kt:25]: map to no risk?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/ddv6/PoliceAreasDDv6.kt
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/ddv6/PoliceAreasDDv6.kt:56]: Here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/Disability.kt [hunk 6]
```diff
@@ -23,5 +23,7 @@ enum class Disability(val value: String) {
       .mapNotNull { disabilityName ->
         entries.find { it.name == disabilityName }?.value
       }
+
+    fun from(value: String?): Disability? = entries.firstOrNull { it.value == value }
```
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/enums/Disability.kt]: No listed condition

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsRetrieveDWandMO.kt [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    version.deviceWearerResponsibleAdult = deviceWearer.toResponsibleAdult(versionId)
+    version.mappa = deviceWearer.toMappa(versionId)
+    version.addresses = listOfNotNull(
+      deviceWearer.toAddress(versionId),
+      deviceWearer.toSecondaryAddress(versionId),
+      deviceWearer.toTertiaryAddress(versionId),
+    ).toMutableList()
+
+    // Do we need to put offence on installation and risk? Maybe for old DD versions?
```
Comment (james-francis-MT) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsRetrieveDWandMO.kt]: Outstanding question

Comment (CslBing) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsRetrieveDWandMO.kt]: When we create new version it should be always the latest DD version

**Review by CslBing — DISMISSED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsDates.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  val dateTimeFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
+
+  fun getBritishDate(dateTime: ZonedDateTime?): String? =
+    dateTime?.toInstant()?.atZone(londonTimeZone)?.format(dateFormatter)
+
+  fun getBritishDateAndTime(dateTime: ZonedDateTime?): String? =
+    dateTime?.toInstant()?.atZone(londonTimeZone)?.format(dateTimeFormatter)
+
+  fun parseDate(date: String): ZonedDateTime =
+    LocalDate.parse(date, dateFormatter).atStartOfDay().atZone(londonTimeZone)
```
Comment (wcdkj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsDates.kt:21]: Haven't followed this all the way through but could we guard against ""? 
"" will throw a DateTimeParseException

Comment (james-francis-MT) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsDates.kt:21]: Any date that could be empty  should be using the parse date or null method. I think this will need a good amount of testing either way as I'm not sure which values are nullable / could be empty. I only had a basic example payload to work with

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsRetrieveDWandMO.kt (hunk 7 above)
Comment (james-francis-MT) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringcreateanorderapi/models/fms/FmsRetrieveDWandMO.kt]: So no need to add offence to installation and risk as they now live in the offences object

**Review by CslBing — DISMISSED** [review]

**Review by CslBing — APPROVED** [review]

**Review by morsh3d — APPROVED** [review]
